src/com/amazon/corretto/crypto/provider/HkdfSecretKeyFactorySpi.java

```diff
@@ -0,0 +1,166 @@
+            spec.info,
+            spec.info.length);
+        break;
+      default:
+        throw new AssertionError(
+            "This should not be reachable, since the constructor of HkdfSpec ensures the value is"
+                + " one of the above choices.");
+    }
+
+    return new SecretKeySpec(resultBytes, spec.algorithmName);
+  }
+
+  private void checkExpandLength(final long outLen) {
+    final long upperLimit = ((long) digestLength) * 255L;
+    if (outLen >= upperLimit) {
+      throw new IllegalArgumentException("Desired output length is too large.");
+    }
+  }
+
+  private static native void hkdf(
+      byte[] jOutput,
+      int outputLen,
+      int digestCode,
+      byte[] jSecret,
+      int secretLen,
+      byte[] jSalt,
+      int saltLen,
+      byte[] jInfo,
+      int infoLen);
+
+  private static native void hkdfExtract(
+      byte[] jOutput,
+      int outputLen,
+      int digestCode,
+      byte[] jSecret,
+      int secretLen,
+      byte[] jSalt,
+      int saltLen);
+
+  private static native void hkdfExpand(
+      byte[] jOutput,
+      int outputLen,
+      int digestCode,
+      byte[] jPrk,
+      int prkLen,
+      byte[] jInfo,
+      int infoLen);
+
+  private static native int getDigestLen(int digestCode);
+
+  @Override
+  protected KeySpec engineGetKeySpec(final SecretKey key, final Class<?> keySpec) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  protected SecretKey engineTranslateKey(final SecretKey key) {
+    throw new UnsupportedOperationException();
+  }
+
+  static final Map<String, HkdfSecretKeyFactorySpi> INSTANCES = getInstances();
+
+  private static final String HKDF = "Hkdf";
+  private static final String WITH = "With";
+  static final String HKDF_WITH_SHA1 = HKDF + WITH + "SHA-1";
+  static final String HKDF_WITH_SHA256 = HKDF + WITH + "SHA-256";
+  static final String HKDF_WITH_SHA384 = HKDF + WITH + "SHA-384";
+  static final String HKDF_WITH_SHA512 = HKDF + WITH + "SHA-512";
+
+  private static final int SHA1_CODE = 1;
+  private static final int SHA256_CODE = 2;
+  private static final int SHA384_CODE = 3;
+  private static final int SHA512_CODE = 4;
+
+  private static Map<String, HkdfSecretKeyFactorySpi> getInstances() {
+    final Map<String, HkdfSecretKeyFactorySpi> result = new HashMap<>();
+    result.put(
+        getSpiFactoryForAlgName(HKDF_WITH_SHA1),
+        new HkdfSecretKeyFactorySpi(SHA1_CODE, getDigestLen(SHA1_CODE)));
+    result.put(
+        getSpiFactoryForAlgName(HKDF_WITH_SHA256),
+        new HkdfSecretKeyFactorySpi(SHA256_CODE, getDigestLen(SHA256_CODE)));
+    result.put(
+        getSpiFactoryForAlgName(HKDF_WITH_SHA384),
+        new HkdfSecretKeyFactorySpi(SHA384_CODE, getDigestLen(SHA384_CODE)));
+    result.put(
+        getSpiFactoryForAlgName(HKDF_WITH_SHA512),
+        new HkdfSecretKeyFactorySpi(SHA512_CODE, getDigestLen(SHA512_CODE)));
+    return Collections.unmodifiableMap(result);
```

Comment:
nit: not a big deal, but this method is a bit redundant as we already [have Utils for getting + caching digest len](https://github.com/corretto/amazon-corretto-crypto-provider/blob/main/src/com/amazon/corretto/crypto/provider/Utils.java#L54-L95). example for going from name to length [here](https://github.com/corretto/amazon-corretto-crypto-provider/blob/74ac74160e426dc973edc0a0cffb108a1f8f7baf/src/com/amazon/corretto/crypto/provider/EvpHmac.java#L316).

it looks like this only gets called once per digest algorithm at JVM start, so the caching aspect isn't very important.

Reply:
Thanks for pointing this out. I removed the redundant code and used the APIs in Utils class instead.